sparse/_coo/numba_extension.py

```diff
@@ -0,0 +1,131 @@
+"""
+Numba support for COO objects.
+
+For now, this just supports attribute access
+"""
+import numpy as np
+import numba
+from numba.extending import (
+    models, register_model, box, unbox, NativeValue, make_attribute_wrapper,
+    overload_method, lower_getattr, type_callable, lower_builtin,
+    overload_attribute
+)
+from numba.targets.imputils import impl_ret_borrowed, lower_constant
+from numba.typing.typeof import typeof_impl
+from numba import cgutils, types, from_dtype
+from llvmlite import ir
+
+from . import COO
+
+__all__ = ['COOType']
+
+
+class COOType(types.Type):
+    def __init__(self, data_dtype: np.dtype, coords_dtype: np.dtype, ndim: int):
+        assert isinstance(data_dtype, np.dtype)
+        assert isinstance(coords_dtype, np.dtype)
+        self.data_dtype = data_dtype
+        self.coords_dtype = coords_dtype
+        self.ndim = ndim
+        super().__init__(name='COOType[{!r}, {!r}, {!r}]'.format(
+        	numba.from_dtype(data_dtype),
+        	numba.from_dtype(coords_dtype),
+        	ndim
+        ))
+
+    @property
+    def key(self):
+        return self.data_dtype, self.coords_dtype, self.ndim
+
+    @property
+    def data_type(self):
+        return numba.from_dtype(self.data_dtype)[:]
+
+    @property
+    def coords_type(self):
+        return numba.from_dtype(self.coords_dtype)[:, :]
+
+    @property
+    def shape_type(self):
+       	return types.UniTuple(types.int64, self.ndim)
```

Comment:
Is there any way to avoid putting `ndim` in the type itself here, or is this a reasonable approach?

Reply:
I think this is reasonable, with a quick search in Numba's codebase, I couldn't find any way to do this without specifying the length.

Reply:
The length is part of the `Tuple` type. `types.UniTuple(types.int64, 1)` is a different type to `types.UniTuple(types.int64, 2)`.